kanidm_proto/src/oauth2.rs

```diff
@@ -0,0 +1,87 @@
+use url::Url;
+use webauthn_rs::base64_data::Base64UrlSafeData;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub enum CodeChallengeMethod {
+    // default to plain if not requested as S256. Reject the auth?
+    // plain
+    // BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+    S256,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AuthorisationRequest {
+    // Must be "code". (or token, see 4.2.1)
+    pub response_type: String,
+    pub client_id: String,
+    pub state: Base64UrlSafeData,
+    // base64?
+    pub code_challenge: Base64UrlSafeData,
+    // Probably also should be an enum.
+    pub code_challenge_method: CodeChallengeMethod,
+    // Uri?
+    pub redirect_uri: Url,
+    // appears to be + seperated?
+    pub scope: String,
+}
+
+/// We ask our user to consent to this Authorisation Request with the
+/// following data.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ConsentRequest {
```

Comment:
While we are sending a response to the user yes, the intent of the response is to request their consent to allow the authorisation to proceed.

So this would probably be an "AuthorisationRequestResponse" if we were to match the http flow.